channelnotifier/channelnotifier.go

```diff
@@ -0,0 +1,192 @@
+
+	ntfnServer *subscribe.Server
+
+	started uint32
+	stopped uint32
+	wg      sync.WaitGroup
+	quit    chan struct{}
+}
+
+// ActiveChannelEvent represents a new event where a channel becomes active.
+type ActiveChannelEvent struct {
+	// Channel is the channel that has become active.
+	Channel *channeldb.OpenChannel
+}
+
+// InactiveChannelEvent represents a new event where a channel becomes inactive.
+type InactiveChannelEvent struct {
+	// Channel is the channel that has become inactive.
+	Channel *channeldb.OpenChannel
+}
+
+// ClosedChannelEvent represents a new event where a channel becomes closed.
+type ClosedChannelEvent struct {
+	// CloseSummary is the summary of the channel close that has occurred.
+	CloseSummary *channeldb.ChannelCloseSummary
+}
+
+// New creates a new channel notifier. The ChannelNotifier gets channel
+// events from peers and from the chain arbitrator, and dispatches them to
+// its clients.
+func New() *ChannelNotifier {
+	return &ChannelNotifier{
+		NewActiveChannels:   queue.NewConcurrentQueue(20),
+		NewInactiveChannels: queue.NewConcurrentQueue(20),
+		ClosedChanEvents:    queue.NewConcurrentQueue(20),
+		ntfnServer:          subscribe.NewServer(),
+		quit:                make(chan struct{}),
+	}
+}
+
+// Start starts the ChannelNotifier and all goroutines it needs to carry out its task.
+func (c *ChannelNotifier) Start() error {
+	if !atomic.CompareAndSwapUint32(&c.started, 0, 1) {
+		return nil
+	}
+
+	log.Tracef("ChannelNotifier %v starting", c)
+
+	if err := c.ntfnServer.Start(); err != nil {
+		return err
+	}
+
+	c.NewActiveChannels.Start()
+	c.NewInactiveChannels.Start()
+	c.ClosedChanEvents.Start()
+
+	c.wg.Add(1)
+	go c.channelEventNotifier()
+
+	return nil
+}
+
+// Stop signals the notifier for a graceful shutdown.
+func (c *ChannelNotifier) Stop() {
+	if !atomic.CompareAndSwapUint32(&c.stopped, 0, 1) {
+		return
+	}
+	close(c.quit)
+
+	c.ntfnServer.Stop()
+
+	c.NewActiveChannels.Stop()
+	c.NewInactiveChannels.Stop()
+	c.ClosedChanEvents.Stop()
+
+	c.wg.Wait()
+}
+
+// SubscribeChannelEvents returns a subscribe.Client that will receive updates
+// any time the Server is made aware of a new event.
+func (c *ChannelNotifier) SubscribeChannelEvents() (*subscribe.Client, error) {
+	return c.ntfnServer.Subscribe()
+}
+
+// NotifyActiveChannelEvent notifies the channelEventNotifier goroutine that a
+// channel has become active.
+func (c *ChannelNotifier) NotifyActiveChannelEvent(channel *channeldb.OpenChannel) {
+	select {
+	case c.NewActiveChannels.ChanIn() <- channel:
+	case <-c.quit:
+	default:
+		log.Warnf("unable to notify on active channel event: active " +
+			"channels queue unexpectedly closed before notifier quit")
+	}
+}
+
+// NotifyInactiveChannelEvent notifies the channelEventNotifier goroutine that a
+// channel has become inactive.
+func (c *ChannelNotifier) NotifyInactiveChannelEvent(channel *channeldb.OpenChannel) {
+	select {
+	case c.NewInactiveChannels.ChanIn() <- channel:
```

Comment:
Using these concurrent queues we risk that events don't reach clients in the order they are notified. Instead we should share the queue for all events, or just send the event directly to the `channelEventNotifier` here.

Reply:
good point, fixed